hashintel/hash
# Add `createUser` GQL mutation

## 🌟 What is the purpose of this PR?




This PR introduces a `createUser` GQL mutation, which can be used by HASH instance admins to create partially/fully signed up users.

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1203007126736613/1203249971975774/f) _(internal)_

## 🔍 What does this change?




See commits.

## 📜 Does this require a change to the docs?



Not that I'm aware of.

## ⚠️ Known issues



- It is not currently possible to create a user and programmatically set their email address to verified.

## 🐾 Next steps



## 🛡 What tests cover this?



None, the GQL API mutation was tested manually.

## ❓ How to test this?



1. Checkout the branch and run the HASH App locally
2. Go to localhost:3000 and login as a HASH instance admin
3. Access the GQL playground at `localhost:5001/graphql`
4. Run following query, and inspect the response to see the created user entity
```GraphQL
mutation createUser($email: UserEmailInput!){
  createUser(
    email: $email,
    password: "password",
    isInstanceAdmin: true,
    shortname: "test",
    preferredName: "Test",
    linkResolveDepth: 5
    linkTargetEntityResolveDepth: 5
  ) {
    roots
  }
}
```
```JSON
{
  "email": {
    "address": "test@example.com",
    "verified": true
  }
}
```
5. Log out of the HASH app (or just clear your browser cookies) and try logging in as the created user at `localhost:3000/login`
